examples/statemachine/src/cli/generator.ts

```diff
@@ -0,0 +1,206 @@
+        });
+        classBodyPublic.append('}', NL);
+
+        for (const event of ctx.statemachine.events) {
+            classBodyPublic.append(NL);
+            classBodyPublic.append(`virtual void ${event.name}() {`, NL);
+            classBodyPublic.indent(methodBody => {
+                methodBody.append('std::cout << "Impossible event for the current state." << std::endl;', NL);
+            });
+            classBodyPublic.append('}', NL);
+        }
+    });
+    ctx.fileNode.append('};', NL);
+}
+
+function generateStatemachineClass(ctx: GeneratorContext) {
+    ctx.fileNode.append(`class ${ctx.statemachine.name} {`, NL);
+    ctx.fileNode.append('private:', NL);
+    ctx.fileNode.indent(classBodyPrivate => {
+        classBodyPrivate.append('State* state = nullptr;', NL);
+    });
+    ctx.fileNode.append(NL);
+
+    ctx.fileNode.append('public:', NL);
+    ctx.fileNode.indent(classBodyPublic => {
+        classBodyPublic.append(`${ctx.statemachine.name}(State* initial_state) {`, NL);
+        classBodyPublic.indent(ctorBody => {
+            ctorBody.append('initial_state->set_context(this);', NL);
+            ctorBody.append('state = initial_state;', NL);
+            ctorBody.append('std::cout << "[" << state->get_name() << "]" << std::endl;', NL);
+        });
+        classBodyPublic.append('}', NL, NL);
+
+        classBodyPublic.append(`~${ctx.statemachine.name}() {`, NL);
+        classBodyPublic.indent(dctorBody => {
+            dctorBody.append('if (state != nullptr) {', NL);
+            dctorBody.indent(thenBody => {
+                thenBody.append('delete state;', NL);
+            });
+            dctorBody.append('}', NL);
+        });
+        classBodyPublic.append('}', NL, NL);
+
+        classBodyPublic.append('void transition_to(State *new_state) {', NL);
+        classBodyPublic.indent(methodBody => {
+            methodBody.append('std::cout << state->get_name() << " ===> " << new_state->get_name() << std::endl;', NL);
+            methodBody.append('if (state != nullptr) {', NL);
+            methodBody.indent(thenBody => {
+                thenBody.append('delete state;', NL);
+            });
+            methodBody.append('}', NL);
+
+            methodBody.append('new_state->set_context(this);', NL);
+
+            methodBody.append('state = new_state;', NL);
+
+        });
+        classBodyPublic.append('}', NL);
+
+        for (const event of ctx.statemachine.events) {
+            classBodyPublic.append(NL);
+            classBodyPublic.append(`void ${event.name}() {`, NL);
+            classBodyPublic.indent(methodBody => {
+                methodBody.append(`state->${event.name}();`, NL);
+            });
+            classBodyPublic.append('}', NL);
+        }
+    });
+    ctx.fileNode.append('};', NL);
+}
+
+function generateStateDeclaration(ctx: GeneratorContext, state: State) {
+    ctx.fileNode.append(`class ${state.name} : public State {`, NL);
```

Comment:
I think it's quite unusual to model the states as class instances. A far more efficient implementation is to define states as integer constants, then the state machine is a big `switch` statement over the current state.

That's not very important, though. We can leave the current implementation for now, or you can refactor it in case you're motivated for it.

Reply:
> I think it's quite unusual to model the states as class instances.
I've implemented the programming pattern [state](https://en.wikipedia.org/wiki/State_pattern).

The approach is very useful for the case, when your program has some states and can react different ways on the same events, depends on the current state. This is exactly what we have: traffic light colors are states, `next` and `switchCapacity` are events.

However, the current version of the `Trafficlight` statemachine only switches between states and do no specific actions, what makes using the state pattern overhead. Anyway, I propose to leave it, because when we add such "specific actions" in the grammar [what will make the example more interesting], it will be extremely easily to update C++ code generation.